src/Lithnet.GoogleApps/CourseTeacherRequestFactory.cs

```diff
@@ -0,0 +1,401 @@
+        {
+            try
+            {
+                this.WaitForGate();
+
+                using (PoolItem<ClassroomService> poolService = this.classRoomServicePool.Take(NullValueHandling.Ignore))
+                {
+                    CoursesResource.TeachersResource.DeleteRequest request = poolService.Item.Courses.Teachers.Delete(courseId, userId);
+                    Trace.WriteLine($"Removing teacher {userId} from course {courseId}");
+                    try
+                    {
+                        request.ExecuteWithRetry(RetryEvents.Backoff | RetryEvents.Timeout, RetryCount);
+                    }
+                    catch (GoogleApiException e)
+                    {
+                        if (!throwOnMissingTeacher && this.IsMissingTeacherError(e.HttpStatusCode, e.Message))
+                        {
+                            return;
+                        }
+
+                        throw;
+                    }
+                }
+            }
+            finally
+            {
+                this.ReleaseGate();
+            }
+        }
+
+        public void AddTeachers(string courseId, IList<Teacher> teachers, bool throwOnExistingTeacher)
+        {
+            if (CourseTeacherRequestFactory.BatchSize <= 1)
+            {
+                foreach (Teacher teacher in teachers)
+                {
+                    this.AddTeacher(courseId, teacher, throwOnExistingTeacher);
+                }
+
+                return;
+            }
+
+            try
+            {
+                this.WaitForGate();
+
+                using (PoolItem<ClassroomService> poolService = this.classRoomServicePool.Take(NullValueHandling.Ignore))
+                {
+                    List<ClientServiceRequest<Teacher>> requests = new List<ClientServiceRequest<Teacher>>();
+
+                    foreach (Teacher teacher in teachers)
+                    {
+                        Trace.WriteLine($"Queuing batch teacher add for {teacher.UserId} to course {courseId}");
+                        requests.Add(poolService.Item.Courses.Teachers.Create(teacher, courseId));
+                    }
+
+                    this.ProcessBatches<Teacher>(courseId, !throwOnExistingTeacher, false, requests, poolService, (batchHelper) =>
+                    {
+                        return (content, error, i, message) =>
+                        {
+                            int index = batchHelper.BaseCount + i;
+                            this.ProcessTeacherResponse(courseId, (teachers[index] as Teacher).UserId, batchHelper.IgnoreExistingTeacher, batchHelper.IgnoreMissingTeacher, error, message, batchHelper.RequestsToRetry, batchHelper.Request, batchHelper.FailedTeachers, batchHelper.Failures);
+                        };
+                    });
+                }
+            }
+            finally
+            {
+                this.ReleaseGate();
+            }
+        }
+
+        protected class ProcessBatchHelper<T>
```

Comment:
Another to move into it's own file

Reply:
Done. Moved to file CourseTeacherRequestFactory.ProcessBatchHelper.cs

Class is now protected internal.